gunrock/applications/sssp/sssp_implementation.hxx

```diff
@@ -17,63 +17,90 @@
+  using vertex_t = typename meta_t::vertex_type;
+  using weight_t = typename meta_t::weight_type;
+   
+  weight_t* distances;
+  vertex_t* predecessors;
+  
+  sssp_result_t(
+    weight_t* distances_,
+    vertex_t* predecessors_
+  ) {
+    distances    = distances_;
+    predecessors = predecessors_;
+  }
+};
+
+
+template <typename d_graph_t, typename meta_t>
+struct sssp_problem_t : problem_t<d_graph_t, meta_t> {
 
-  using weight_pointer_t = typename graph_type::weight_pointer_t;
-  using vertex_pointer_t = typename graph_type::vertex_pointer_t;
+  using param_t   = sssp_param_t<meta_t>;
+  using result_t  = sssp_result_t<meta_t>;
 
-  // Useful types from problem_t
-  using problem_type = problem_t<graph_type, host_graph_type>;
+  using vertex_t = typename meta_t::vertex_type;
+  using edge_t   = typename meta_t::edge_type;
+  using weight_t = typename meta_t::weight_type;
 
   vertex_t single_source;
-  weight_pointer_t distances;
-  vertex_pointer_t predecessors;
+  weight_t* distances;
+  vertex_t* predecessors;
+  
   thrust::device_vector<vertex_t> visited;
 
-  /**
-   * @brief Construct a new sssp problem t object
-   *
-   * @param G graph on GPU
-   * @param g graph on CPU
-   * @param context system context
-   * @param source input single source for sssp
-   * @param dist output distance pointer
-   * @param preds output predecessors pointer
-   */
-  sssp_problem_t(graph_type* G,
-                 host_graph_type* g,
+  sssp_problem_t(d_graph_t* d_G,
+                 meta_t* meta,
                  std::shared_ptr<cuda::multi_context_t> context,
-                 vertex_t& source,
-                 weight_pointer_t dist,
-                 vertex_pointer_t preds)
-      : problem_type(G, g, context),
-        single_source(source),
-        distances(dist),
-        predecessors(preds),
-        visited(g[0].get_number_of_vertices(), -1) {
-    // Set all initial distances to INFINITY
-    auto d_dist = thrust::device_pointer_cast(distances);
-    thrust::fill(thrust::device, d_dist + 0,
-                 d_dist + g[0].get_number_of_vertices(),
-                 std::numeric_limits<weight_t>::max());
-    thrust::fill(thrust::device, d_dist + source, d_dist + source + 1, 0);
+                 param_t& param,
+                 result_t& result)
+      : problem_t<d_graph_t, meta_t>(d_G, meta, context) {
+    
+    single_source = param.single_source;
+    distances     = result.distances;
+    predecessors  = result.predecessors;
+    
+    auto n_vertices = meta[0].get_number_of_vertices();
+    
+    auto d_distances = thrust::device_pointer_cast(distances);
+    thrust::fill(
+      thrust::device,
+      d_distances + 0,
+      d_distances + n_vertices,
+      std::numeric_limits<weight_t>::max()
+    );
+    thrust::fill(thrust::device, d_distances + single_source, d_distances + single_source + 1, 0);
+    
+    visited.resize(n_vertices);
+    thrust::fill(thrust::device, visited.begin(), visited.end(), -1);
```

Comment:
This is not a resize, this should be done within the constructor and should not even use a `fill`, it should initialize to `-1`, like how it was before. 
```cpp
sssp_problem_t(d_graph_t* d_G,
                 meta_t* meta,
                 std::shared_ptr<cuda::multi_context_t> context,
                 ...)
      : : problem_t<d_graph_t, meta_t>(d_G, meta, context),
            ...
        visited(g[0].get_number_of_vertices(), -1) {
// ... other constructor stuff.
}
```